pkg/cloudclient/gcp/private.go

```diff
@@ -0,0 +1,199 @@
+		Rrdatas: svcIPs,
+		Type:    "A",
+		Ttl:     30,
+	}
+
+	clusterDNS := &configv1.DNS{}
+	err = kclient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, clusterDNS)
+	if err != nil {
+		return err
+	}
+
+	var zones []configv1.DNSZone
+	if clusterDNS.Spec.PublicZone != nil {
+		zones = append(zones, *clusterDNS.Spec.PublicZone)
+	}
+	if clusterDNS.Spec.PrivateZone != nil {
+		zones = append(zones, *clusterDNS.Spec.PrivateZone)
+	}
+
+	for _, zone := range zones {
+		// Look for an existing resource record set in the zone.
+		listCall := c.dnsService.ResourceRecordSets.List(c.projectID, zone.ID)
+		response, err := listCall.Name(newRRSet.Name).Do()
+		if err != nil {
+			return err
+		}
+
+		dnsChange := &gdnsv1.Change{
+			Additions: []*gdnsv1.ResourceRecordSet{newRRSet},
+		}
+
+		// There will be at most one result but loop anyway.
+		// An empty slice will proceed directly to Create.
+		for _, rrset := range response.Rrsets {
+			if !equalResourceRecordSets(newRRSet, rrset) {
+				dnsChange.Deletions = append(dnsChange.Deletions, rrset)
+			}
+		}
+
+		changesCall := c.dnsService.Changes.Create(c.projectID, zone.ID, dnsChange)
+		_, err = changesCall.Do()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) removeDNSForService(kclient client.Client, svc *corev1.Service, dnsName string) error {
+	// google.golang.org/api/dns/v1.Service is a struct, not an interface, which
+	// will make this all but impossible to write unit tests for
+
+	svcIPs, err := getIPAddressesFromService(svc)
+	if err != nil {
+		return err
+	}
+
+	baseDomain, err := baseutils.GetClusterBaseDomain(kclient)
+	if err != nil {
+		return err
+	}
+
+	dnsChange := &gdnsv1.Change{
+		Deletions: []*gdnsv1.ResourceRecordSet{
+			{
+				Name:    dnsName + "." + baseDomain + ".",
+				Rrdatas: svcIPs,
+				Type:    "A",
+				Ttl:     30,
+			},
+		},
+	}
+
+	clusterDNS := &configv1.DNS{}
+	err = kclient.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, clusterDNS)
+	if err != nil {
+		return err
+	}
+
+	var zones []configv1.DNSZone
+	if clusterDNS.Spec.PublicZone != nil {
+		zones = append(zones, *clusterDNS.Spec.PublicZone)
+	}
+	if clusterDNS.Spec.PrivateZone != nil {
+		zones = append(zones, *clusterDNS.Spec.PrivateZone)
+	}
+
+	for _, zone := range zones {
+		call := c.dnsService.Changes.Create(c.projectID, zone.ID, dnsChange)
+		_, err = call.Do()
+		if err != nil {
+			dnsError, ok := err.(*googleapi.Error)
+			if !ok || dnsError.Code != http.StatusNotFound {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func getIPAddressesFromService(svc *corev1.Service) ([]string, error) {
+	var ips []string
+	for _, ingress := range svc.Status.LoadBalancer.Ingress {
+		ips = append(ips, ingress.IP)
+	}
+
+	if len(ips) == 0 {
+		return nil, cioerrors.NewLoadBalancerNotReadyError()
+	}
+
+	return ips, nil
+}
+
+func equalResourceRecordSets(rrset1, rrset2 *gdnsv1.ResourceRecordSet) bool {
+	// Just compare the fields we set ourselves.
```

Comment:
For clarity, maybe add a comment that if any of these expressions are false, then you're returning false. We only return true if every expression is true.

Reply:
Agreed.  I'd like to play with it a bit more and see if `reflect.DeepEqual` could work here.  I haven't yet observed GCP altering or adding info in these structs.

Reply:
Got `reflect.DeepEqual` working on the record comparisons.